File: 算法/代码随想录算法PDF/00调试源代码/107.二叉树的层序遍历-ii.cpp
/*
 * @lc app=leetcode.cn id=107 lang=cpp
 *
 * [107] 二叉树的层序遍历 II
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        queue<TreeNode*> que;
        vector<vector<int>> res;
        if (root) que.push(root);
        while (!que.empty()) {
            int size = que.size();
            vector<int> layer;
            while (size) {
                TreeNode* node = que.front();
                que.pop();
                layer.push_back(node->val);
                if (node->left) que.push(node->left);
                if (node->right) que.push(node->right);
                size--;
            }
            res.push_back(layer);
        }
        reverse(res.begin(), res.end());
        return res;
    }
};
// @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/107.二叉树的层序遍历-ii.py
#
# @lc app=leetcode.cn id=107 lang=python3
#
# [107] 二叉树的层序遍历 II
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrderBottom(self, root: Optional[TreeNode]) -> List[List[int]]:
        from collections import deque
        que = deque([])
        res = []
        if root:
            que.append(root)
        while que:
            layer = []
            size = len(que)
            while size:
                node = que.popleft()
                layer.append(node.val)
                if node.left:
                    que.append(node.left)
                if node.right:
                    que.append(node.right)
                size -= 1
            res.append(layer)
        return res[::-1]
# @lc code=end

